File: ethereumj-core/src/main/java/org/ethereum/net/p2p/PeersMessage.java
/*
 * The MIT License (MIT)
 *
 * Copyright 2017 Alexander Orlov <alexander.orlov@loxal.net>. All rights reserved.
 * Copyright (c) [2016] [ <ether.camp> ]
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

package org.ethereum.net.p2p;

import org.ethereum.util.ByteUtil;
import org.ethereum.util.RLP;
import org.ethereum.util.RLPList;
import org.spongycastle.util.encoders.Hex;

import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

/**
 * Wrapper around an Ethereum Peers message on the network
 *
 * @see org.ethereum.net.p2p.P2pMessageCodes#PEERS
 */
public class PeersMessage extends P2pMessage {

    private boolean parsed = false;

    private Set<Peer> peers;

    public PeersMessage(final byte[] payload) {
        super(payload);
    }

    public PeersMessage(final Set<Peer> peers) {
        this.peers = peers;
        parsed = true;
    }

    private void parse() {
        final RLPList paramsList = (RLPList) RLP.decode2(encoded).get(0);

        peers = new LinkedHashSet<>();
        for (int i = 1; i < paramsList.size(); ++i) {
            final RLPList peerParams = (RLPList) paramsList.get(i);
            final byte[] ipBytes = peerParams.get(0).getRLPData();
            final byte[] portBytes = peerParams.get(1).getRLPData();
            final byte[] peerIdRaw = peerParams.get(2).getRLPData();

            try {
                final int peerPort = ByteUtil.byteArrayToInt(portBytes);
                final InetAddress address = InetAddress.getByAddress(ipBytes);

                final String peerId = peerIdRaw == null ? "" : Hex.toHexString(peerIdRaw);
                final Peer peer = new Peer(address, peerPort, peerId);
                peers.add(peer);
            } catch (final UnknownHostException e) {
                throw new RuntimeException("Malformed ip", e);
            }
        }
        this.parsed = true;
    }

    private void encode() {
        final byte[][] encodedByteArrays = new byte[this.peers.size() + 1][];
        encodedByteArrays[0] = RLP.encodeByte(this.getCommand().asByte());
        final List<Peer> peerList = new ArrayList<>(this.peers);
        for (int i = 0; i < peerList.size(); i++) {
            encodedByteArrays[i + 1] = peerList.get(i).getEncoded();
        }
        this.encoded = RLP.encodeList(encodedByteArrays);
    }

    @Override
    public byte[] getEncoded() {
        if (encoded == null) encode();
        return encoded;
    }

    public Set<Peer> getPeers() {
        if (!parsed) this.parse();
        return peers;
    }

    @Override
    public P2pMessageCodes getCommand() {
        return P2pMessageCodes.PEERS;
    }

    @Override
    public Class<?> getAnswerMessage() {
        return null;
    }

    public String toString() {
        if (!parsed) this.parse();

        final StringBuilder sb = new StringBuilder();
        for (final Peer peerData : peers) {
            sb.append("\n       ").append(peerData);
        }
        return "[" + this.getCommand().name() + sb.toString() + "]";
    }
}

File: free-ethereum-core/src/main/java/org/ethereum/net/p2p/PeersMessage.kt
/*
 * The MIT License (MIT)
 *
 * Copyright 2017 Alexander Orlov <alexander.orlov@loxal.net>. All rights reserved.
 * Copyright (c) [2016] [ <ether.camp> ]
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

package org.ethereum.net.p2p

import org.ethereum.util.ByteUtil
import org.ethereum.util.RLP
import org.ethereum.util.RLPList
import org.spongycastle.util.encoders.Hex
import java.net.InetAddress
import java.net.UnknownHostException
import java.util.*

/**
 * Wrapper around an Ethereum Peers message on the network

 * @see org.ethereum.net.p2p.P2pMessageCodes.PEERS
 */
class PeersMessage : P2pMessage {

    private var peers: MutableSet<Peer>? = null

    constructor(payload: ByteArray) : super(payload)

    constructor(peers: MutableSet<Peer>) {
        this.peers = peers
        parsed = true
    }

    private fun parse() {
        val paramsList = RLP.decode2(encoded)[0] as RLPList

        peers = LinkedHashSet<Peer>()
        for (i in 1..paramsList.size - 1) {
            val peerParams = paramsList[i] as RLPList
            val ipBytes = peerParams[0].rlpData
            val portBytes = peerParams[1].rlpData
            val peerIdRaw = peerParams[2].rlpData

            try {
                val peerPort = ByteUtil.byteArrayToInt(portBytes)
                val address = InetAddress.getByAddress(ipBytes)

                val peerId = if (peerIdRaw == null) "" else Hex.toHexString(peerIdRaw)
                val peer = Peer(address, peerPort, peerId)
                peers!!.add(peer)
            } catch (e: UnknownHostException) {
                throw RuntimeException("Malformed ip", e)
            }

        }
        this.parsed = true
    }

    private fun encode() {
        val encodedByteArrays = arrayOfNulls<ByteArray>(this.peers!!.size + 1)
        encodedByteArrays[0] = RLP.encodeByte(this.command.asByte())
        val peerList = ArrayList(this.peers!!)
        for (i in peerList.indices) {
            encodedByteArrays[i + 1] = peerList[i].encoded
        }
        this.encoded = RLP.encodeList(*encodedByteArrays)
    }

    override fun getEncoded(): ByteArray {
        if (encoded == null) encode()
        return encoded
    }

    fun getPeers(): Set<Peer> {
        if (!parsed) this.parse()
        return peers!!
    }

    override fun getCommand(): P2pMessageCodes {
        return P2pMessageCodes.PEERS
    }

    override fun getAnswerMessage(): Class<*>? {
        return null
    }

    override fun toString(): String {
        if (!parsed) this.parse()

        val sb = StringBuilder()
        for (peerData in peers!!) {
            sb.append("\n       ").append(peerData)
        }
        return "[" + this.command.name + sb.toString() + "]"
    }
}